src/tokens/ERC1155.vy

```diff
@@ -111,65 +180,88 @@ def supportsInterface(interface_id: bytes4) -> bool:
+        owned the token.
     @param to The 20-byte receiver address.
     @param id The 32-byte identifier of the token.
     @param amount The 32-byte token amount that is
-           being transferred.
+        being transferred.
     @param data The maximum 1024-byte additional data
-           with no specified format.
+        with no specified format.
     """
-    pass
+    assert owner == msg.sender or self.isApprovedForAll[owner][msg.sender], "ERC1155: Not authorized to transfer"
+    assert to != empty(address), "ERC1155: transfer to the zero address"
+    assert self._balances[owner][id] >= amount, "ERC1155: insufficient balance"
+
+    self._balances[owner][id] = unsafe_sub(self._balances[owner][id], amount)
+    self._balances[to][id] = unsafe_add(self._balances[to][id], amount)
+
+    log TransferSingle(msg.sender, owner, to, id, amount)
+    assert self._check_on_erc1155_received(owner, to, id, amount, data), "ERC1155: transfer to non-ERC1155Receiver implementer"
 
 
 @external
-def safeBatchTransferFrom(owner: address, to: address, ids: DynArray[uint256, max_value(uint16)], amounts: DynArray[uint256, max_value(uint16)],
-                          data: Bytes[1024]):
+def safeBatchTransferFrom(
+    owner: address,
+    to: address,
+    ids: DynArray[uint256, max_value(uint16)],
+    amounts: DynArray[uint256, max_value(uint16)],
+    data: Bytes[1024]
+):
     """
     @dev Batched version of `safeTransferFrom`.
     @notice Note that `ids` and `amounts` must have the
-            same length. Also, if `to` refers to a smart
-            contract, it must implement {IERC1155Receiver-onERC1155Received}
-            and return the acceptance magic value.
+        same length. Also, if `to` refers to a smart
+        contract, it must implement {IERC1155Receiver-onERC1155Received}
+        and return the acceptance magic value.
     @param owner The 20-byte address which previously
-           owned the token.
+        owned the token.
     @param to The 20-byte receiver address.
     @param ids The 32-byte array of token identifiers. Note
-           that the order and length must match the 32-byte
-           `amounts` array.
+        that the order and length must match the 32-byte
+        `amounts` array.
     @param amounts The 32-byte array of token amounts that are being
-           transferred. Note that the order and length must match
-           the 32-byte `ids` array.
+        transferred. Note that the order and length must match
+        the 32-byte `ids` array.
     @param data The maximum 1024-byte additional data
-           with no specified format.
+        with no specified format.
     """
-    pass
+    assert owner == msg.sender or self.isApprovedForAll[owner][msg.sender], "ERC1155: Not authorized to transfer"
+    assert to != empty(address), "ERC1155: transfer to the zero address"
+    assert len(ids) == len(amounts), "ERC1155: batch lengths mismatch"
+
+    for i in range(max_value(uint16)):
```

Comment:
we can do the following since `ids` is iterable if I'm correct:
```vyper
for id in ids:
    ...
```

Reply:
`ids` is iterable, but since both `ids` and `amounts` have to be indexed by the same index, this seemed like the best approach (unless vyper has something like enumerate).

Reply:
Unfortunately, Vyper does not support `enumerate` currently. I, fortunately, opened a VIP already some time ago :-D https://github.com/vyperlang/vyper/issues/3080. IMHO the more appropriate solution would be that track the index ourselves. I.e.

```vyper
idx: uint256 = 0

for id in ids:
        amount: uint256 = amounts[idx]
        assert self._balances[owner][id] >= amount, "ERC1155: insufficient balance"
        self._balances[owner][id] = unsafe_sub(self._balances[owner][id], amount)
        self._balances[to][id] = unsafe_add(self._balances[to][id], amount)
        unsafe_add(idx, 1)
```

Like that we can save the gas for the checked increment arithmetic & if you use `range(max_value(uint16))`, `max_value(uint16)` will be the STOP condition even though maybe `ids` are lower. So we would need to use the length here that we need to cache before in order to save gas. So self-tracking is the most efficient & clean version. What do you think?

Reply:
This makes sense. Will update!